### DISCUSSION TITLE
CYD code for displaying photo from web

### DISCUSSION THREAD

**Discussion by webhooki** [witnessmenow/ESP32-Cheap-Yellow-Display#82] [Q&A]
Yea so like in the title. Im struggling to write or find working code for just displaying jpg from url on the screen. Did someone done that?

**Comment by witnessmenow** [comment]
I wrote a library for fetching files off the internet that will hopefully help. 

https://github.com/witnessmenow/file-fetcher-arduino

This example should do what most of what your looking for, other than its a png rather jpg

https://github.com/witnessmenow/file-fetcher-arduino/blob/main/examples/displayImageToLCD/displayImageToLCD.ino

Here is a different example of displaying a JPG, so if you use the jpg displaying part of this code, with the above example, 

https://github.com/witnessmenow/Spotify-Diy-Thing/blob/main/SpotifyDiyThing/cheapYellowLCD.h#L177

**Reply by webhooki** [reply]
First - Great library. Thank you very muchh. Second - what are the requirments for photo in the url? Your example url works like a charm, but i tried a lot of photos and none of them are displayed. (png, exact dimensions, small file, from lots of sites, even from this from example.) Only your url works. Sometimes there are zero errors, sometimes error with code 6. What can be the issue? But still thats exactly what i was searching, good job.

**Reply by witnessmenow** [reply]
Hmm, I don't know that actually!

Can you share an example url that doesn't work?

**Reply by webhooki** [reply]
hmmm. This for example. https://drive.google.com/uc?export=view&id=1ov9dCi18CLM78uo1GfQBUMtopHxtCjhH
exact dimensions, format etc. Or this https://upload.wikimedia.org/wikipedia/commons/b/bd/Prostor-320x240.png
or this https://i.scdn.co/image/ab67616d00004851cebdf1f7660ace8c2a80585c (but this is a jpg from spotify api)

**Comment by witnessmenow** [comment]
The Google drive link may have some extra steps like redirection etc, or maybe the fact it doesn't have a file ending as part of the url is a problem, not actually sure

The Spotify thumbnail definitely should work, as that's what the second example I linked to is doing, and the file fetcher code is based on what I originally did for the Spotify thumbnail code

**Comment by dave0003** [comment]
I'm trying to use your example code, saveImageToMemory, to fetch an image and display it.  I'm having some trouble extracting the relevant parts of the displayImageToLCD example to do the displaying part using the imageFile pointer.  Do you know what I might have to change to get it to work?

Edit:  I got it to compile by changing this:

`int rc = png.open((const char *)imageFileUri, myOpen, myClose, myRead, mySeek, PNGDraw);`
to
`int rc = png.openRAM(imageFile, imageSize, PNGDraw);`

I also had to use the "CYD2USB" version of User_Setup.h.

I am able to get it to work if the URL is https, but not for http.  For http, I get a 400 Bad Request error, and the log is gibberish.

For example, https:

`192.168.1.100 - - [14/Nov/2024:04:36:01 +0800] "GET /gate.png HTTP/1.0" 200 81221 "-" "Arduino/Arduino"`

http (the gibberish changes with each request):

`192.168.1.100 - - [14/Nov/2024:04:41:58 +0800] "\x16\x03\x03\x01\x1E\x01\x00\x01\x1A\x03\x03\x00\x00\x00\x1B\x93\xAF\xE3\xF9^.\x15\xB8\xC6\xDFE\xA0\x12,\xD1emQV;\xC5\x85\xBE\x00\xA1\x94\x9F\xD5\x00\x00\xA2\xC0,\xC00\xC0\xAD\xC0$\xC0(\xC0" 400 157 "-" "-"`

**Comment by dave0003** [comment]
Here's what I ended up with, to fetch a 320x240 jpeg from my webcam and display it.  It loops repeatedly, and gets about 1-2 fps.  Right now it only works with HTTPS.  It requires the attached saveImageToMemory.txt (change to .ino, it's modified from the example code), and FileFetcher.cpp and FileFetcher.h (no changes).

[saveImageToMemory.txt](https://github.com/user-attachments/files/17740741/saveImageToMemory.txt)

**Comment by clach04** [comment]
@webhooki  is this worth marking as answered now?
